edk2toollib/uefi/edk2/path_utilities.py

```diff
@@ -19,64 +19,91 @@
                 else:
                     # assume current working dir relative.  Will catch invalid dir when checking whole list
-                    self.PackagePathList.append(os.path.abspath(os.path.join(os.getcwd(), a)))
+                    candidate_package_path_list.append(Path(os.getcwd(), a))
 
         error = False
-        for a in self.PackagePathList[:]:
-            if (not os.path.isdir(a)):
-                self.logger.log(logging.ERROR if error_on_invalid_pp else logging.WARNING,
-                                "Invalid package path entry {0}".format(a))
-                self.PackagePathList.remove(a)  # remove invalid path
+        for a in candidate_package_path_list[:]:
+            if not a.is_dir():
+                self.logger.log(logging.ERROR if error_on_invalid_pp else
+                                logging.WARNING,
+                                f"Invalid package path entry {a.resolve()}")
+                candidate_package_path_list.remove(a)
                 error = True
 
-        # report error
-        if (error and error_on_invalid_pp):
-            raise Exception("Invalid package path directory(s)")
-
-        # for each package path, trace from packagepath to the
-        # either the workspace root or filesystem root and verify
-        # no *.DEC file exists. This would signify a nested package.
-        for package_path in self.PackagePathList:
-            p = Path(package_path)
-            ws = self.WorkspacePath
-            while p != p.parent:
-                if str(p).lower() == str(ws).lower():
-                    break
-                if len(glob.glob(f'{p}/*dec')) != 0:
-                    raise Exception(f'Nested packages not allowed. Pkg path [{package_path}] nested in Package [{p}]')
-                p = p.parent
+        self.PackagePathList = [str(p) for p in candidate_package_path_list]
+
+        if error and error_on_invalid_pp:
+            raise NotADirectoryError(errno.ENOENT, os.strerror(errno.ENOENT),
+                                     a.resolve())
+
+        # Nested package check - ensure packages do not exist in a linear
+        # path hierarchy.
+        package_path_packages = {}
+        for package_path in candidate_package_path_list:
+            package_path_packages[package_path] = \
+                [Path(p).parent for p in package_path.glob('**/*.dec')]
+
+        for package_path, packages in package_path_packages.items():
+            for i, package in enumerate(packages):
+                for j in range(i + 1, len(packages)):
+                    comp_package = packages[j]
+
+                    if (package.is_relative_to(comp_package)
+                            or comp_package.is_relative_to(package)):
+                        raise Exception(
+                            f"Nested packages not allowed. The packages "
+                            f"[{str(package)}] and [{str(comp_package)}] are "
+                            f"nested")
```

Comment:
would appreciate a sentence or two more about this algorithm for checking for nesting.  What the intermediate  lists/dictionaries are tracking and used for.  Requires too much reverse engineering to understand.

Reply:
Added in PR Update commit https://github.com/tianocore/edk2-pytool-library/pull/186/commits/be6ddc562fdc981aaefb31ef83e4b1e47510ab38